### ISSUE TITLE
Support for using multiple identity materials and trust materials not working

### ISSUE THREAD

**Issue by patpatpat123** [Hakky54/sslcontext-kickstart#58]
Hello,

First of all, a big thanks for this project.
It is well written, well documented, and very helpful!

I am opening an issue regarding the "Support for using multiple identity materials and trust materials" as I believe it might not be behaving correctly.

In order to illustrate my example, I am going to simplify to two different services (instead of a large number)

First service I need to talk to: OrderService. It is a service where one will place orders. (I order a video game, I order a bicycle)

Second service I need to take to, DeliveryService. It is a service where one will deliver something to someone. It can work a standalone, I deliver this contract to this person, often used in conjunction of OrderService, I deliver the video game I just ordered to someone.

The two companies owning the services are competing company, they will enforce client to have it their own keystores. No way possible to go to one company and say: If I bring you the certificate I use to talk to the other service, can you trust it as well? No no.

In order to perform such, I am using this library, where I beforehand verified.

```
 SSLFactory sslFactory = SSLFactory.builder()
                    .withIdentityMaterial(Paths.get(/path/to/orderservice/client-to-talk-to-orderservice-keystore.p12), "talkToOrderService".toCharArray())
                    .withTrustMaterial(Paths.get(/path/to/truststore.p12), "trustStorePassPhrase".toCharArray())
                    .build();
            return NettySslUtils.forClient(sslFactory).build();
```

This is working fine to talk to order service

But also, I verified,

```
 SSLFactory sslFactory = SSLFactory.builder()
                    .withIdentityMaterial(Paths.get(/path/to/deliveryservice/client-to-talk-to-deliveryservice-keystore.p12), "talkToDeliveryService".toCharArray())
                    .withTrustMaterial(Paths.get(/path/to/truststore.p12), "trustStorePassPhrase".toCharArray())
                    .build();
            return NettySslUtils.forClient(sslFactory).build();
```

This is working fine to talk to delivery service (mock the step to talk to order service)

Now, chaining the two calls, first talk to order service, then to delivery service.

```
 SSLFactory sslFactory = SSLFactory.builder()
                    .withIdentityMaterial(Paths.get(/path/to/orderservice/client-to-talk-to-orderservice-keystore.p12), "talkToOrderService".toCharArray())
                    .withIdentityMaterial(Paths.get(/path/to/deliveryservice/client-to-talk-to-deliveryservice-keystore.p12), "talkToDeliveryService".toCharArray())
                    .withTrustMaterial(Paths.get(/path/to/truststore.p12), "trustStorePassPhrase".toCharArray())
                    .build();
            return NettySslUtils.forClient(sslFactory).build();
```

This still yields ```javax.net.ssl.SSLHandshakeException: Received fatal alert: unknown_ca; nested exception is io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: Received fatal alert: unknown_ca```

Could you please help?

Thank you

**Comment by Hakky54** [comment]
Hi @patpatpat123 

Thank you for this detailed description!

If you get an alert unknown_ca back from the server, then the server did not like the certificate you've send as the client certificate, because it is not signed by a CA which is trusted by the server for client certificates. Well you have tested both certificates separately and that worked so the certificates are definitely correct! However I think the underlying KeyManager has trouble sending the correct client certificate to the server when having two client certificates.

I have added an integration for this use case for netty http client, see here: https://github.com/Hakky54/sslcontext-kickstart/blob/36b1a944c4de377e09a08cb3960ca92e28d855d0/sslcontext-kickstart-for-netty/src/test/java/nl/altindag/ssl/SSLFactoryIT.java#L79
The test works but it is currently disabled because of the server rate limiter. But if you run this on your environment it should also work.

So I am concluding that Netty is capable of handling this kind of SSL configuration, of having multiple client certificates within a single SslContext. So I need some more information to better analyse your issue. Could you set a breakpoint on the following statements:
https://github.com/Hakky54/sslcontext-kickstart/blob/36b1a944c4de377e09a08cb3960ca92e28d855d0/sslcontext-kickstart/src/main/java/nl/altindag/ssl/keymanager/CompositeX509ExtendedKeyManager.java#L81
and 
https://github.com/Hakky54/sslcontext-kickstart/blob/36b1a944c4de377e09a08cb3960ca92e28d855d0/sslcontext-kickstart/src/main/java/nl/altindag/ssl/keymanager/CompositeX509ExtendedKeyManager.java#L96
And share here what the parameter values are of k **String[] keyTypes** and **Principal[] issuers**

You don't need to share the exact values of the **Principal[] issuers** as it can be confidential. But can you tell me if the array is empty, null or filled?

**Comment by patpatpat123** [comment]
Many thanks for the detailed steps.

I cloned the project and started to play with the SSLFactoryIT class.

I first split executeHttpsRequestWithMutualAuthenticationForMultipleClientIdentitiesWithSingleSslConfiguration, adding two more tests.

```
 @Test
    void testOne() throws IOException {
        SSLFactory sslFactory = SSLFactory.builder()
                .withIdentityMaterial(Paths.get("/path/to/order-service.keystore"), "password".toCharArray())
                .withTrustMaterial(Paths.get("/path/to/order-service-truststore.p12"), "password".toCharArray())
                .build();

        SslContext sslContext = NettySslUtils.forClient(sslFactory).build();
        HttpClient httpClient = HttpClient.create().secure(sslSpec -> sslSpec.sslContext(sslContext));

        Tuple2<String, Integer> responseEntity2 = httpClient.post().uri("https://order-service:443/order/123")
                .responseSingle((response, body) -> Mono.zip(body.asString(), Mono.just(response.status().code()))).block();

```

This works ✅
Then adding 

```
@Test
    void testTwo() throws IOException {
        SSLFactory sslFactory = SSLFactory.builder()
                 .withIdentityMaterial(Paths.get("/path/to/delivery-service.p12"), "password".toCharArray())
                .withTrustMaterial(Paths.get("/path/to/delivery-service-truststore.p12"), "password".toCharArray())
                .build();

        SslContext sslContext = NettySslUtils.forClient(sslFactory).build();
        HttpClient httpClient = HttpClient.create().secure(sslSpec -> sslSpec.sslContext(sslContext));

        Tuple2<String, Integer> responseEntity = httpClient.post().uri("https://delivery-service/delivery/1/1/dlivery/")
                .responseSingle((response, body) -> Mono.zip(body.asString(), Mono.just(response.status().code()))).block();

```
This works ✅

Finally, the initial test with the two combined

```
 @Test
    void testThree() throws IOException {
        LogCaptor logCaptor = LogCaptor.forName("nl.altindag.ssl");

        SSLFactory sslFactory = SSLFactory.builder()
  .withIdentityMaterial(Paths.get("/path/to/order-service.keystore"), "password".toCharArray())
   .withIdentityMaterial(Paths.get("/path/to/delivery-service.p12"), "password".toCharArray())
                .withTrustMaterial(Paths.get("/path/to/order-service-truststore.p12"), "password".toCharArray())
                .withIdentityMaterial(Paths.get("/path/to/delivery-service-truststore.p12"), "password".toCharArray())
                .build();

        SslContext sslContext = NettySslUtils.forClient(sslFactory).build();
        HttpClient httpClient = HttpClient.create().secure(sslSpec -> sslSpec.sslContext(sslContext));

        Tuple2<String, Integer> responseEntity2 = httpClient.post().uri("https://order-service:443/order/123")
                .responseSingle((response, body) -> Mono.zip(body.asString(), Mono.just(response.status().code())))
                .block();

        Tuple2<String, Integer> responseEntity = httpClient.post().uri("https://delivery-service/delivery/1/1/delivery/")
                .responseSingle((response, body) -> Mono.zip(body.asString(), Mono.just(response.status().code())))
                .block();
```
This failed ‼️with the unknown CA issue.

**Comment by patpatpat123** [comment]
I followed your advice and start adding breakpoints to CompositeX509ExtendedKeyManager line 81 and 96.

When I run the first two tests, it does not go to breakpoints (probably expected)

When I run the third test (the one with multiple) it first break at line 96.

There: String[] keyTypes is ["RSA"]
**issuers is null**

And I try inverting the orders of the files in SSLFactory sslFactory = SSLFactory.builder(), and inverting the http call order.

I believe to have tried all combinations .withIdentityMaterial(order-service).withIdentityMaterial(delivery-service) + http call to order then to delivery

.withIdentityMaterial(order-service).withIdentityMaterial(delivery-service) + http call to delivery then to order

.withIdentityMaterial(delivery-service).withIdentityMaterial(order-service) + http call to order then to delivery

.withIdentityMaterial(delivery-service).withIdentityMaterial(order-service) + http call to delivery then to order

for all cases, the keyTypes is [RSA]
**the issuer is always null**

May I ask what is the issue please?
Or please tell me any further information needed.

Thank you

**Comment by Hakky54** [comment]
Thank you very much for testing all the scenarios! Your test result confirms my initial thoughts about this use case. I will try to explain what should be happening, what is happening and how I can try to fix this issue.

The KeyManager and the underlying methods are responsible for selecting the right key material to proceed the ssl handshake. The starting point within the java code of selecting the right key material happens from the client side at the method `chooseClientAlias` or `chooseEngineClientAlias` they use the supplied parameters `keyTypes`, `issuers`, `socket` and `sslEngine` to select the correct client alias. It will grab and always return the first matching one based on the parameters. When the client alias has been obtained it will proceed to get the certificate chain and private key.

In a traditional setup the KeyManager instance has most of the time just one keypair. It can easily get an alias even for example if the issuer is empty/null. However when there are multiple keypairs within the KeyManager instance, in this case the `CompositeX509ExtendedKeyManager` has multiple keypairs, the keymanager will just grab the first keypair (the client certificate) which is matching with the supported keyTypes. So if both keypairs support RSA it will just grab the first matching one and use that one to communicate with the server because it cannot do an additional filtering based on the issuers because it is null or empty. In the integration test which I just added has the issuer present and therefor it is able to correctly select the correct client alias. But as we have seen in your case the issuers can also be null.

I think this library is missing the capability of handling that use case. I think I am able to find a possible solution that might do the trick and should work for you. Please let me know if you wanna wait and test for it. I think it would be ready by tomorrow end of the day.

**Comment by Hakky54** [comment]
And can you share how you created the keypair and jks/p12 file? In that way I will be able to repoduce the same behaviour what you are getting.

**Comment by patpatpat123** [comment]
More than happy to wait and test once ready!

As for the key, sadly, the were created by someone else.
I have the password though, hence can extract and run commands, just let me know what you need! 🙂

**Comment by Hakky54** [comment]
During my testing I discovered that the certificates which I created with java `keytool` have always the `issuers` present within the KeyManager method `chooseEngineClientAlias`. But the certificates which I have created with openssl sometimes have the issuers. I don't know what the exact reason is, it could be that I didn't created the certificates correctly with openssl. But I think these certificates are still valid certificates and therefor It should be somehow still be supported when using multiple identities for example with your use case.

I have created an initial implementation of the solution. It is currently a work-in-progress, so method names and variables can change.

We have concluded that the keymanager has not enough information to decide which certificate to use and therefore it is just getting the first matching alias from the method `chooseEngineClientAlias`. In your case it does not match. The root cause is that `issuers` is null or empty. A solution would be to recreate the certificates, but you are not able to do that.

An alternative would be giving just enough information to the underlying KeyManager instance to decide which certificate it should use for which downstream api. This configuration will ensure you to have a single http client with a single ssl configuration that will work with multiple downstream api's.

The solution which I currently implemented can have an additional parameter for the SSLFactory which takes the specific certificate alias in the keystore/p12 file as a key and should have a host as a value parameter. For example: `some-alias` -> `https://localhost:8443/api/hello`. What you are basically saying to the KeyManager is that if your client is trying to communicate with server on localhost with port 8443 it should use the alias `some-alias`.

Could you retry your tests with the following setup:
```
SSLFactory sslFactory = SSLFactory.builder()
        .withIdentityMaterial(Paths.get("/path/to/order-service.keystore"), "password".toCharArray())
        .withIdentityMaterial(Paths.get("/path/to/delivery-service.p12"), "password".toCharArray())
        .withTrustMaterial(Paths.get("/path/to/order-service-truststore.p12"), "password".toCharArray())
        .withClientIdentityRoute("order-service-certificate-alias", "https://host-of-order-service:port-of-order-server")
        .withClientIdentityRoute("delivery-service-certificate-alias", "https://host-of-delivery-service:port-of-delivery-server")
        .build();
```

These changes are available on a different branch, so you can try it out on the following branche: https://github.com/Hakky54/sslcontext-kickstart/pull/61

**Comment by patpatpat123** [comment]
Many thanks for this PR!

Unfortunately, the issue still persists. 
But maybe I am doing things wrong!

So what I did:
I cloned the branch, and built it on my local.
I then imported the local built jar, and rebuilt my business services. Saw the new API .withClientIdentityRoute() valid. Was red before.

I input the two routes, the two alias, as they are in the client certificates.
Verified, the two aliases are different.
However, still getting the unknown_ca issue.

Could you please suggest what can I do to debug, or to gather further information?
Will be glad to paste what I can here

**Comment by Hakky54** [comment]
Hmm strange,

What might be helpful is to share the full logs with your 3 test cases which you mentioned previously while having the VM arguments present: `-Djavax.net.debug=SSL,keymanager,trustmanager,ssl:handshake`

**Comment by Hakky54** [comment]
By the way, what we also can do is a remote video session. In that way I will be able to debug and analyse your issue better. Let me know if you wanna try that out

**Comment by Hakky54** [comment]
We had a remote session together with Patrick and went though his code snippet with the solution provided earlier. He had couple of small issues with the configuration, but after resolving that it worked. He was able to have a single ssl configuration for to use with a single http client to communicate with n amount of servers. The configuration is described here: [Routing client identity material to specific host](https://github.com/Hakky54/sslcontext-kickstart#routing-identity-material-to-specific-host)

This option is available from version 6.3.0 onwards
